apps/src/templates/certificates/petition/PetitionForm.jsx

```diff
@@ -0,0 +1,107 @@
+import React, {useState} from 'react';
+import {Button} from 'react-bootstrap';
+import {range, mapValues} from 'lodash';
+import {
+  createErrorMessage,
+  keyValidation
+} from '@cdo/apps/templates/certificates/petition/petitionHelpers';
+import ControlledFieldGroup from '@cdo/apps/templates/certificates/petition/ControlledFieldGroup';
+
+const PetitionForm = () => {
+  // data starts with all required fields having an empty value to ensure proper validation
+  const [data, setData] = useState(mapValues(keyValidation, () => ''));
+  const [errorMessage, setErrorMessage] = useState('');
+
+  const handleChange = e => {
+    e.persist();
+    setData(data => ({...data, [e.target.name]: e.target.value}));
+  };
+
+  const handleSubmit = e => {
+    e.preventDefault();
+    setErrorMessage(createErrorMessage(data));
+  };
+
+  const buildControlledFieldGroup = (
+    id,
+    placeholderOrLabel,
+    helpText,
+    componentClass,
+    children
+  ) => (
+    <ControlledFieldGroup
+      id={id}
+      placeholderOrLabel={placeholderOrLabel}
+      helpText={helpText}
+      componentClass={componentClass}
+      onChange={handleChange}
+      value={data[id] || ''}
+    >
+      {children}
+    </ControlledFieldGroup>
+  );
+
+  return (
+    <>
+      <form
+        id="petition-form"
+        className="petition-form"
+        onSubmit={handleSubmit}
+      >
+        <div className={'petition-space'}>{errorMessage}</div>
+        {buildControlledFieldGroup('name', 'Name')}
```

Comment:
Ya, those components were exactly what I was thinking! I was hoping to get rid of these 'build' functions also by doing that, tho. Imo they're a little hard to parse when they have a bunch of similar arguments without the argument names (and so putting a component in there tells us what the arguments are)

Reply:
Ah got it. I think that'll work too––will investigate 🤓

Reply:
Hmm, here's my problem: I need to put that component (maybe like PetitionFieldGroup) outside of PetitionForm so it isn't rendered thousands of time. But if I do that, I lose access to the handleChange function ...https://stackoverflow.com/questions/42573017/in-react-es6-why-does-the-input-field-lose-focus-after-typing-a-character

Reply:
oh, i see! you should be able to wrap your `handleChange` and `handleSubmit` in `useCallback` to prevent them from being redefined every render